File: bin4j/Format.java
package bin4j;

import java.nio.ByteBuffer;

import static bin4j.Pair.pair;
import java.io.UnsupportedEncodingException;

public abstract class Format<T> implements XFunction<T, ByteBuffer>
{
    public static Format<String> string = new Format<String>()
    {
        public ByteBuffer apply(String s)
        {
            byte[] bytes;
            try
            {
                bytes = s.getBytes("US-ASCII");
            }
            catch (UnsupportedEncodingException e)
            {
                throw new RuntimeException(e);
            }
            ByteBuffer result = ByteBuffer.allocate(4 + bytes.length);
            result.putInt(bytes.length);
            result.put(bytes);
            result.position(0);
            return result;
        }        

        public String unapply(ByteBuffer b)
        {
            int length = b.getInt();
            byte[] bytes = new byte[length];
            b.get(bytes);
            try
            {
                return new String(bytes, "US-ASCII");
            }
            catch (UnsupportedEncodingException e)
            {
                throw new RuntimeException(e);
            }
        }
    };

    public <U> Format2<T, U> andThen(final Format<U> other)
    {
        return new Format2<T, U>()
        {
            public ByteBuffer apply(T t, U u)
            {
                return ByteBuffers.sequence(Format.this.apply(t), other.apply(u));
            }

            public Pair<T, U> unapply(ByteBuffer b)
            {
                return pair(Format.this.unapply(b), other.unapply(b));
            }
        };
    }

    public <U> Format2<T, U> bind(final Function<T, Format<U>> uf)
    {
        return new Format2<T, U>()
        {
            public ByteBuffer apply(T t, U u)
            {
                Format<U> fu = uf.apply(t);
                return ByteBuffers.sequence(Format.this.apply(t), fu.apply(u));
            }

            public Pair<T, U> unapply(ByteBuffer b)
            {
                T t = Format.this.unapply(b);
                Format<U> fu = uf.apply(t);
                return pair(t, fu.unapply(b));
            }                
        };
    }

    public <U> Format<U> map(final XFunction<T, U> xmap)
    {
        return new Format<U>()
        {
            public ByteBuffer apply(U u)
            {
                return Format.this.apply(xmap.unapply(u));
            }

            public U unapply(ByteBuffer b)
            {
                return xmap.apply(Format.this.unapply(b));
            }
        };
    }
}

File: src/main/java/org/bin4s/Format.scala
package org.bin4s

import java.nio.ByteBuffer
import java.io.UnsupportedEncodingException

abstract class Format[T] extends XFunction[T, ByteBuffer] {
  def andThen[U](other: Format[U]): Format[(T, U)] = new Format[(T, U)] {
    def apply(tu: (T, U)): ByteBuffer = tu match { case (t, u) => ByteBuffers.sequence(Format.this.apply(t), other.apply(u)) }
    def unapply(b: ByteBuffer): (T, U) = (Format.this.unapply(b), other.unapply(b))
  }

  def bind[U](uf: T => Format[U]): Format[(T, U)] = new Format[(T, U)] {
    def apply(tu: (T, U)): ByteBuffer = {
      val fu: Format[U] = uf.apply(tu._1)
      ByteBuffers.sequence(Format.this.apply(tu._1), fu.apply(tu._2))
    }

    def unapply(b: ByteBuffer): (T, U) = {
      val t: T = Format.this.unapply(b)
      val fu: Format[U] = uf.apply(t)
      (t, fu.unapply(b))
    }
  }

  def map[U](xmap: XFunction[T, U]): Format[U] = new Format[U] {
    def apply(u: U): ByteBuffer = Format.this.apply(xmap.unapply(u))
    def unapply(b: ByteBuffer): U = xmap.apply(Format.this.unapply(b))
  }
}

object Format {
  var string: Format[String] = new Format[String] {
    def apply(s: String): ByteBuffer = {
      var bytes: Array[Byte] = null
      try {
        bytes = s.getBytes("US-ASCII")
      }
      catch {
        case e: UnsupportedEncodingException => {
          throw new RuntimeException(e)
        }
      }
      val result: ByteBuffer = ByteBuffer.allocate(4 + bytes.length)
      result.putInt(bytes.length)
      result.put(bytes)
      result.position(0)
      result
    }

    def unapply(b: ByteBuffer): String = {
      val length: Int = b.getInt
      val bytes: Array[Byte] = new Array[Byte](length)
      b.get(bytes)
      try
        new String(bytes, "US-ASCII")
      catch {
        case e: UnsupportedEncodingException => {
          throw new RuntimeException(e)
        }
      }
    }
  }
}
